bson/bsoncodec/registry_examples_test.go

```diff
@@ -99,27 +117,177 @@ func ExampleRegistry_customDecoder() {
 				return err
 			}
 			result = i64 != 0
-		case bsontype.Double:
-			f64, err := vr.ReadDouble()
+		default:
+			return fmt.Errorf(
+				"received invalid BSON type to decode into lenientBool: %s",
+				vr.Type())
+		}
+
+		val.SetBool(result)
+		return nil
+	}
+
+	reg := bson.NewRegistry()
+	reg.RegisterTypeDecoder(
+		lenientBoolType,
+		bsoncodec.ValueDecoderFunc(lenientBoolDecoder))
+
+	// Marshal a BSON document with a single field "isOK" that is a non-zero
+	// integer value.
+	b, err := bson.Marshal(bson.M{"isOK": 1})
+	if err != nil {
+		panic(err)
+	}
+
+	// Now try to decode the BSON document to a struct with a field "IsOK" that
+	// is type lenientBool. Expect that the non-zero integer value is decoded
+	// as boolean true.
+	type MyDocument struct {
+		IsOK lenientBool `bson:"isOK"`
+	}
+	var doc MyDocument
+	err = bson.UnmarshalWithRegistry(reg, b, &doc)
+	if err != nil {
+		panic(err)
+	}
+	fmt.Printf("%+v\n", doc)
+	// Output: {IsOK:true}
+}
+
+func ExampleRegistry_RegisterKindEncoder() {
+	// Create a custom encoder that writes any Go type that has underlying type
+	// int32 as an a BSON int64. To do that, we register the encoder as a "kind"
+	// encoder for kind reflect.Int32. That way, even user-defined types with
+	// underlying type int32 will be encoded as a BSON int64.
+	int32To64Encoder := func(
+		ec bsoncodec.EncodeContext,
+		vw bsonrw.ValueWriter,
+		val reflect.Value,
+	) error {
+		// All encoder implementations should check that val is valid and is of
+		// the correct type or kind before proceeding.
+		if !val.IsValid() || val.Kind() != reflect.Int32 {
+			return bsoncodec.ValueEncoderError{
+				Name:     "int32To64Encoder",
+				Kinds:    []reflect.Kind{reflect.Int32},
+				Received: val,
+			}
+		}
+
+		return vw.WriteInt64(val.Int())
+	}
+
+	// Create a default registry and register our int32-to-int64 encoder for
+	// kind reflect.Int32.
+	reg := bson.NewRegistry()
+	reg.RegisterKindEncoder(
+		reflect.Int32,
+		bsoncodec.ValueEncoderFunc(int32To64Encoder))
+
+	// Define a document that includes an int32, an int64, and a user-defined
+	// type "myInt" that has underlying type int32.
```

Comment:
```suggestion
	// type "myInt" that aliases int32.
```

Reply:
A type alias and a type definition with underlying type are different things. For clarity, a type alias is something like `type myInt = int32` and a type definition with an underlying type is something like `type myInt int32`.

A type alias is not actually a new type but is just a new name for the same type; type aliases only exist in code (i.e. there are no references in the compiled binary). A type definition with underlying type is a new type and exists both in code and as a distinct type in the compiled binary. Check out an example that shows the different behavior of both [here](https://go.dev/play/p/L5OufN4d5P7).

In this case, `RegisterKindEncoder` is specifically used to handle any types with the same underlying type. Type aliases can be handled by `RegisterTypeEncoder` because they are actually all the same type.